### PR TITLE
New card charts 

### LINKED ISSUE #3356: [Card Chart] New Card charts
### What package is this for?

- [x] React
- [ ] Angular

### Summary

We need a new set of cards with charts for MAS Core 8.8 release, such as:

**Meter chart**
image

**Sparkline chart**
image

**Stacked bar chart**
image

**Stacked area chart**
image

### Specific timeline issues / requests

Do you want this work within a specific time period? Yes, we need it before the end of May.
Is it related to an upcoming release? Yes, MAS Core 8.8

### You will consider this component complete when...

- [ ] It's available as an export of this library

### Available extra resources

What resources do you have to assist this effort? I can work on it

### PR DESCRIPTION
**Pull request by cgirani**
Closes #3356 

**Summary**

- New card charts for `MeterChart/ SparklineChart (AreaChart)/ StackedAreaChart`

**Change List (commits, features, bugs, etc)**

- feat(charts): new sparkline and meter chart
- feat(stackedareachart): stackedareachartcard new component
- chore(export): export new charts and chart cards component
 - feat(chartcard): support to threshold in chart cards
 - chore(snapshot): snapshot update

**Acceptance Test (how to verify the PR)**

- Make sure the new charts works fine

**Regression Test (how to make sure this PR doesn't break old functionality)**

- New code should not affect existing code.

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

**Comment by bjornalm** [comment]
Seeing some clipped text at the bottom in this empty story
image

**Comment by bjornalm** [comment]
RTL mode needs some attention
image

**Comment by bjornalm** [comment]
There are several size and break points combinations that don't work well, I've captured a few of them here. Not sure which combinations that are required to work:
image

File: packages/react/src/components/SparklineChartCard/SparklineChartCard.jsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      toolbar: {
+        enabled: false,
+      },
+    }),
+    [contentWithDefaults]
+  );
+
+  const [listHeight, setListHeight] = useState();
+  useEffect(() => {
+    setListHeight(listRef?.current?.clientHeight);
```
Comment (bjornalm) [inline packages/react/src/components/SparklineChartCard/SparklineChartCard.jsx:102]: Why is this needed?

File: packages/react/src/components/SparklineChartCard/_sparkline-chart-card.scss [hunk 2]
```diff
@@ -0,0 +1,18 @@
+@import '../../globals/vars';
+
+.#{$iot-prefix}--sparkline-card--wrapper {
+  width: 100%;
+  height: calc(100% - (var(--card-list-height) * 1.3));
```
Comment (bjornalm) [inline packages/react/src/components/SparklineChartCard/_sparkline-chart-card.scss:5]: what does the number 1.3 represent?

**Comment by bjornalm** [comment]
A bit of RTL adjustments needed here as well
image

**Comment by bjornalm** [comment]
There are problems with the size breakpoints for this chart as well
image

**Review by bjornalm — CHANGES_REQUESTED** [review]
I haven't looked at all charts but it seems that we need to address RTL and size breakpoint combinations

**Comment by cgirani** [comment]
> Seeing some clipped text at the bottom in this empty story image

Nica catch, it was actually a card issue, I will add now a story for the empty card state with footer ;)

**Comment by cgirani** [comment]
> RTL mode needs some attention 

Checking with carbon charts team about it.

File: packages/react/src/components/SparklineChartCard/SparklineChartCard.jsx (hunk 1 above)
Comment (cgirani) [reply packages/react/src/components/SparklineChartCard/SparklineChartCard.jsx:102]: The list under the chart doesnt have a default height set, so I thought about having this to know the height the chart can take, I would expect a list with max 3/4 items
image
image

File: packages/react/src/components/MeterChartCard/MeterChartCard.story.jsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import { layout05 } from '@carbon/layout';
+
+import { CARD_SIZES } from '../../constants/LayoutConstants';
+import { getCardMinSize } from '../../utils/componentUtilityFunctions';
+
+import MeterChartCardREADME from './MeterChartCard.mdx';
+import MeterChartCard from './MeterChartCard';
+
+export default {
+  title: '1 - Watson IoT/MeterChartCard',
```
Comment (davidicus) [inline packages/react/src/components/MeterChartCard/MeterChartCard.story.jsx:13]: These stories for the cards should be nested inside the Card story. 
```suggestion
  title: '1 - Watson IoT/Card/MeterChartCard',
```

File: packages/react/src/components/StackedAreaChartCard/StackedAreaChart.story.jsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+import { layout05 } from '@carbon/layout';
+
+import { CARD_SIZES } from '../../constants/LayoutConstants';
+import { getCardMinSize } from '../../utils/componentUtilityFunctions';
+
+import StackedAreaChartCardREADME from './StackedAreaChartCard.mdx';
+import StackedAreaChartCard from './StackedAreaChartCard';
+
+export default {
+  title: '1 - Watson IoT/StackedAreaChartCard',
```
Comment (davidicus) [inline packages/react/src/components/StackedAreaChartCard/StackedAreaChart.story.jsx:13]:
```suggestion
  title: '1 - Watson IoT/Card/StackedAreaChartCard',
```

File: packages/react/src/components/MeterChartCard/MeterChartCard.jsx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        enabled: false,
+      },
+    }),
+    [contentWithDefaults, values]
+  );
+  const newSize = useMemo(() => getUpdatedCardSize(size), [size]);
+  const previousChartData = usePrevious(values);
+
+  /** This is needed to update the chart when the lines and values change */
+  useEffect(() => {
```
Comment (davidicus) [inline packages/react/src/components/MeterChartCard/MeterChartCard.jsx:116]: @cgirani Is it possible to bring these effects and size update functions into the base Card component? If we are going to add a prop type to Card in the future to dictate the type of chart we want the card to be able to handle all of these merges and size updates.

File: packages/react/src/components/MeterChartCard/MeterChartCard.jsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+          ],
+        },
+      },
+      toolbar: {
+        enabled: false,
+      },
+    }),
+    [contentWithDefaults, values]
+  );
+  const newSize = useMemo(() => getUpdatedCardSize(size), [size]);
```
Comment (davidicus) [inline packages/react/src/components/MeterChartCard/MeterChartCard.jsx:116]: We run `getUpdatedCardSize` in the Card component. Do we need it here as well?

**Review by bjornalm — APPROVED** [review]
Nice work, 
looks good once Davids comments are addressed.

File: packages/react/src/components/MeterChartCard/MeterChartCard.jsx (hunk 5 above)
Comment (cgirani) [reply packages/react/src/components/MeterChartCard/MeterChartCard.jsx]: it's implemented, having support for Meter/Sparkline/StackedArea [check the commit ](https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3393/commits/1e6cef5e41abb88d95fac05f31a6ce005252d0b1)

File: packages/react/src/components/Card/CardTypeContent.jsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+  const contentWithDefaults = useMemo(() => defaultsDeep({}, content, defaultProps.content), [
+    content,
+  ]);
+
+  const options = getChartOptions(type, data.length > 1, contentWithDefaults);
+
+  useEffect(() => {
+    setListHeight(listRef?.current?.clientHeight);
+  }, [listRef]);
```
Comment (davidicus) [inline packages/react/src/components/Card/CardTypeContent.jsx:46]: Are we adding this ref to just get rid of the linter warning? Just want to be sure we know that changes to refs do not cause a component to re-render. If you want to re-render when the height changes we have to put `listRef.current?.clientHeight` in the dependency array. 
```suggestion
  }, [listRef.current?.clientHeight]);
```

File: packages/react/src/components/Card/CardTypeContent.jsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  const [listHeight, setListHeight] = useState();
+
+  const contentWithDefaults = useMemo(() => defaultsDeep({}, content, defaultProps.content), [
+    content,
+  ]);
+
+  const options = getChartOptions(type, data.length > 1, contentWithDefaults);
+
+  useEffect(() => {
+    setListHeight(listRef?.current?.clientHeight);
```
Comment (davidicus) [inline packages/react/src/components/Card/CardTypeContent.jsx:45]: nit: Since we are defining `listRef` variable on line 35 it does not need the `?` check as we know it exist.

**Review by davidicus — COMMENTED** [review]
It looks good. Just the one question about the dependency array.

**Review by davidicus — APPROVED** [review]
